### ISSUE TITLE
Missing Phosphor dependency

### ISSUE THREAD

**Issue by TomTom4** [jupyter-widgets/ipyleaflet#562]
Hi guys, I am trying to port [ipyleaflet-legend](https://github.com/weatherforce/ipyleaflet-legend) within ipyleaflet as a LegendControl as dicussed [here](https://github.com/weatherforce/ipyleaflet-legend/issues/5). 
So far I have imported ipyleaflet into a virtualenv, made it worked and looked at some notebooks on the example folder.
Then, I went on js folder, yarn install, and  on build, had following errors:
```bash
yarn run v1.22.4
$ webpack
Hash: 4cb96083ede3c8624272240f436b7f4a6dabb7c84fd62e2bd782adb1531f
Version: webpack 4.43.0
Child
    Hash: 4cb96083ede3c8624272
    Time: 123ms
    Built at: 2020-04-28 10:18:38
           Asset      Size  Chunks             Chunk Names
    extension.js  1.09 KiB       0  [emitted]  main
    Entrypoint main = extension.js
    [0] ./src/extension.js 398 bytes {0} [built]
    
    WARNING in configuration
    The 'mode' option has not been set, webpack will fallback to 'production' for this value. Set 'mode' option to 'development' or 'production' to enable defaults for each environment.
    You can also set it to 'none' to disable any default behavior. Learn more: https://webpack.js.org/configuration/mode/
Child
    Hash: 240f436b7f4a6dabb7c8
    Time: 3603ms
    Built at: 2020-04-28 10:18:41
     29 assets
    Entrypoint main = index.js index.js.map
      [1] ./src/layers/Layer.js 4.31 KiB {0} [built]
      [5] ./src/controls/Control.js 1.19 KiB {0} [built]
      [8] ./src/layers/RasterLayer.js 1.11 KiB {0} [built]
     [10] ./src/layers/TileLayer.js 1.88 KiB {0} [built]
     [11] ./src/layers/VectorLayer.js 441 bytes {0} [built]
     [12] ./src/layers/Path.js 1.05 KiB {0} [built]
     [13] ./src/layers/Polyline.js 1.81 KiB {0} [built]
     [14] ./src/layers/Polygon.js 572 bytes {0} [built]
     [15] ./src/layers/CircleMarker.js 862 bytes {0} [built]
     [16] ./src/layers/LayerGroup.js 1.32 KiB {0} [built]
     [17] ./src/layers/FeatureGroup.js 566 bytes {0} [built]
     [19] ./src/notebook.js 294 bytes {0} [built]
     [20] ./src/index.js 565 bytes {0} [built]
     [21] ./src/jupyter-leaflet.js 7.35 KiB {0} [built]
    [119] ./package.json 1.45 KiB {0} [built]
        + 105 hidden modules
    
    WARNING in configuration
    The 'mode' option has not been set, webpack will fallback to 'production' for this value. Set 'mode' option to 'development' or 'production' to enable defaults for each environment.
    You can also set it to 'none' to disable any default behavior. Learn more: https://webpack.js.org/configuration/mode/
    
    ERROR in ./src/controls/WidgetControl.js
    Module not found: Error: Can't resolve '@phosphor/messaging' in '/home/tomtom/weatherforce/Projects/ipyleaflet/js/src/controls'
     @ ./src/controls/WidgetControl.js 7:19-49
     @ ./src/jupyter-leaflet.js
     @ ./src/index.js
     @ ./src/notebook.js
    
    ERROR in ./src/layers/Layer.js
    Module not found: Error: Can't resolve '@phosphor/messaging' in '/home/tomtom/weatherforce/Projects/ipyleaflet/js/src/layers'
     @ ./src/layers/Layer.js 5:19-49
     @ ./src/jupyter-leaflet.js
     @ ./src/index.js
     @ ./src/notebook.js
    
    ERROR in ./src/layers/Popup.js
    Module not found: Error: Can't resolve '@phosphor/messaging' in '/home/tomtom/weatherforce/Projects/ipyleaflet/js/src/layers'
     @ ./src/layers/Popup.js 7:19-49
     @ ./src/jupyter-leaflet.js
     @ ./src/index.js
     @ ./src/notebook.js
    
    ERROR in ./src/controls/WidgetControl.js
    Module not found: Error: Can't resolve '@phosphor/widgets' in '/home/tomtom/weatherforce/Projects/ipyleaflet/js/src/controls'
     @ ./src/controls/WidgetControl.js 8:17-45
     @ ./src/jupyter-leaflet.js
     @ ./src/index.js
     @ ./src/notebook.js
    
    ERROR in ./src/layers/Layer.js
    Module not found: Error: Can't resolve '@phosphor/widgets' in '/home/tomtom/weatherforce/Projects/ipyleaflet/js/src/layers'
     @ ./src/layers/Layer.js 6:17-45
     @ ./src/jupyter-leaflet.js
     @ ./src/index.js
     @ ./src/notebook.js
    
    ERROR in ./src/layers/Popup.js
    Module not found: Error: Can't resolve '@phosphor/widgets' in '/home/tomtom/weatherforce/Projects/ipyleaflet/js/src/layers'
     @ ./src/layers/Popup.js 8:17-45
     @ ./src/jupyter-leaflet.js
     @ ./src/index.js
     @ ./src/notebook.js
Child
    Hash: 4fd62e2bd782adb1531f
    Time: 2679ms
    Built at: 2020-04-28 10:18:41
     29 assets
    Entrypoint main = index.js index.js.map
      [1] ./src/layers/Layer.js 4.31 KiB {0} [built]
      [5] ./src/controls/Control.js 1.19 KiB {0} [built]
      [8] ./src/layers/RasterLayer.js 1.11 KiB {0} [built]
     [10] ./src/layers/TileLayer.js 1.88 KiB {0} [built]
     [11] ./src/layers/VectorLayer.js 441 bytes {0} [built]
     [12] ./src/layers/Path.js 1.05 KiB {0} [built]
     [13] ./src/layers/Polyline.js 1.81 KiB {0} [built]
     [14] ./src/layers/Polygon.js 572 bytes {0} [built]
     [15] ./src/layers/CircleMarker.js 862 bytes {0} [built]
     [16] ./src/layers/LayerGroup.js 1.32 KiB {0} [built]
     [17] ./src/layers/FeatureGroup.js 566 bytes {0} [built]
     [19] ./src/embed.js 144 bytes {0} [built]
     [20] ./src/index.js 565 bytes {0} [built]
     [21] ./src/jupyter-leaflet.js 7.35 KiB {0} [built]
    [119] ./package.json 1.45 KiB {0} [built]
        + 105 hidden modules
    
    WARNING in configuration
    The 'mode' option has not been set, webpack will fallback to 'production' for this value. Set 'mode' option to 'development' or 'production' to enable defaults for each environment.
    You can also set it to 'none' to disable any default behavior. Learn more: https://webpack.js.org/configuration/mode/
    
    ERROR in ./src/controls/WidgetControl.js
    Module not found: Error: Can't resolve '@phosphor/messaging' in '/home/tomtom/weatherforce/Projects/ipyleaflet/js/src/controls'
     @ ./src/controls/WidgetControl.js 7:19-49
     @ ./src/jupyter-leaflet.js
     @ ./src/index.js
     @ ./src/embed.js
    
    ERROR in ./src/layers/Layer.js
    Module not found: Error: Can't resolve '@phosphor/messaging' in '/home/tomtom/weatherforce/Projects/ipyleaflet/js/src/layers'
     @ ./src/layers/Layer.js 5:19-49
     @ ./src/jupyter-leaflet.js
     @ ./src/index.js
     @ ./src/embed.js
    
    ERROR in ./src/layers/Popup.js
    Module not found: Error: Can't resolve '@phosphor/messaging' in '/home/tomtom/weatherforce/Projects/ipyleaflet/js/src/layers'
     @ ./src/layers/Popup.js 7:19-49
     @ ./src/jupyter-leaflet.js
     @ ./src/index.js
     @ ./src/embed.js
    
    ERROR in ./src/controls/WidgetControl.js
    Module not found: Error: Can't resolve '@phosphor/widgets' in '/home/tomtom/weatherforce/Projects/ipyleaflet/js/src/controls'
     @ ./src/controls/WidgetControl.js 8:17-45
     @ ./src/jupyter-leaflet.js
     @ ./src/index.js
     @ ./src/embed.js
    
    ERROR in ./src/layers/Layer.js
    Module not found: Error: Can't resolve '@phosphor/widgets' in '/home/tomtom/weatherforce/Projects/ipyleaflet/js/src/layers'
     @ ./src/layers/Layer.js 6:17-45
     @ ./src/jupyter-leaflet.js
     @ ./src/index.js
     @ ./src/embed.js
    
    ERROR in ./src/layers/Popup.js
    Module not found: Error: Can't resolve '@phosphor/widgets' in '/home/tomtom/weatherforce/Projects/ipyleaflet/js/src/layers'
     @ ./src/layers/Popup.js 8:17-45
     @ ./src/jupyter-leaflet.js
     @ ./src/index.js
     @ ./src/embed.js
error Command failed with exit code 2.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.

``` 
Clearly there is dependencies missing. Looking at ipywidgets, I've seen that it embeds @lumino dependencies and based on [this ipywidget issue](https://github.com/jupyterlab/jupyterlab/issues/7534), I think a good fix would be to put those two (@phosphor/messaging & @phosphor/widgets) as external dependencies in the webpack.config.js file.
Should I make the change and ask for a PR? :)

**Comment by martinRenou** [comment]
Thanks a lot for pointing this out. It seems to be an issue with the switch to Lumino in ipywidgets. We will need to switch to Lumino as well. Making @phosphor/messaging & @phosphor/widgets dependencies will not fix this, we actually need to stop importing stuffs from Phosphor.

**Comment by SylvainCorlay** [comment]
> Thanks a lot for pointing this out. It seems to be an issue with the switch to Lumino in ipywidgets. We will need to switch to Lumino as well. Making @phosphor/messaging & @phosphor/widgets dependencies will not fix this, we actually need to stop importing stuffs from Phosphor.

ipywidgets ha only switched to lumino in master, not in a release, so I don't think this is it.

**Comment by martinRenou** [comment]
@TomTom4 are you using ipywidgets master?

**Comment by TomTom4** [comment]
Wow thanks to answering that fast ! 

>@TomTom4 are you using ipywidgets master?

Nope, I am using the pypi package, which version is 
```bash
ipywidgets==7.5.1
```
But I don't get how webpack could build without giving him the hint that a dependency is either external or listed as a dependency on the package.json? :/ 
by the way, I've tried to list it as external and it built, so tried it out with LayerGroup.ipynb and it doesn't work since it cannot find @phosphor... 
```bash 
[W 11:33:42.547 NotebookApp] 404 GET /static/@phosphor/messaging.js?v=20200428113240 (127.0.0.1) 1.50ms referer=http://localhost:8888/notebooks/LayerGroup.ipynb
[W 11:33:42.548 NotebookApp] 404 GET /static/@phosphor/widgets.js?v=20200428113240 (127.0.0.1) 1.29ms referer=http://localhost:8888/notebooks/LayerGroup.ipynb
```

**Comment by martinRenou** [comment]
> But I don't get how webpack could build without giving him the hint that a dependency is either external or listed as a dependency on the package.json?

You are right, this is wrong and it was a missing dependency. The reason why it worked is because Phosphor is an ipywidgets dependency, but it should be a hard dependency as we are using it directly.

My comment about Lumino is not right as of today.

**Comment by TomTom4** [comment]
>You are right, it was a missing dependency. The reason why it worked is because Phosphor is an ipywidgets dependency, but it should be a hard dependency as we are using it directly.

Okay then ! so how should we proceed? should I make the change on my side, ask for PR, or do you prefer to do it on your side, and then I'll pull from your side? 

Thank you anyway ! :grin:

**Comment by martinRenou** [comment]
> should I make the change on my side, ask for PR

That would be awesome, thanks! I'll merge it right away.

**Comment by martinRenou** [comment]
Fixed by #563 Thanks a lot!
